core/src/main/scala/org/apache/spark/storage/BlockManagerMasterHeartbeatEndpoint.scala

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.storage
+
+import scala.collection.mutable
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.rpc.{RpcCallContext, RpcEnv, ThreadSafeRpcEndpoint}
+import org.apache.spark.storage.BlockManagerMessages.{BlockManagerHeartbeat, RegisterBlockManager, RemoveExecutor, UpdateBlockInfo}
+
+/**
+ * Separate heartbeat out of BlockManagerMasterEndpoint due to performance consideration.
+ */
+private[spark] class BlockManagerMasterHeartbeatEndpoint(
+    override val rpcEnv: RpcEnv,
+    isLocal: Boolean)
+  extends ThreadSafeRpcEndpoint with Logging {
+
+  // Mapping from block manager id to the block manager's information.
+  private val blockManagerLastSeen = new mutable.HashMap[BlockManagerId, Long]
```

Comment:
I know this refactored from original code, but still feel weird that we don't ever use the `lastSeen` elsewhere.

Reply:
Yes, `def lastSeenMs: Long = _lastSeenMs` is never used in original code.

Reply:
So, shall we eliminate `blockManagerLastSeen` as it may occupy a bunch of memory when there's thousands of executors/BlockManagers ?

Reply:
Let me see. Does it have a bug in original code?

Reply:
Yes. we could eliminate `blockManagerLastSeen ` for now at least. The only useful code here is the return value of `def heartbeatReceived`. The lastSeen it stored is no needed.

Reply:
Make sense.